File: C++/022. Generate Parentheses.cpp
class Solution {
private:
void CombinationPar(vector<string>& result, string& sample, int deep, int n, int leftNum, int rightNum)
{
    if (deep==2*n)
    {
        result.push_back(sample);
        return;
    }
    if(leftNum<n)
    {
        sample.push_back('(');
        CombinationPar(result, sample, deep+1, n, leftNum+1, rightNum);
        sample.resize(sample.length()-1);
    }
    if(leftNum>rightNum)
    {
        sample.push_back(')');
        CombinationPar(result, sample, deep+1, n, leftNum, rightNum+1);
        sample.resize(sample.length()-1);
    }
}
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        string sample;
        if(n!=0)
        {
            CombinationPar(res, sample, 0, n, 0, 0);
        }
        return res;
    }
};

File: python/022. Generate Parentheses.py
class Solution(object):
    def dfs(self, n, leftParentheses, cur, res):
        if len(cur) == 2 * n:
            res.append(copy.copy(cur))
            return
        # the number of "(" in cur is leftParentheses + (len(cur) - leftParentheses) / 2
        if leftParentheses + (len(cur) - leftParentheses) / 2 < n:
            cur += "("
            self.dfs(n, leftParentheses+1, cur, res)
            cur = cur[:len(cur)-1]
        if leftParentheses > 0:
            cur += ")"
            self.dfs(n, leftParentheses-1, cur, res)
            cur = cur[:len(cur)-1]
        return
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        cur = ""
        res = []
        self.dfs(n, 0, cur, res)
        return res